lib/apiservers/engine/backends/cache/layer_cache.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"sync"
+)
+
+// LCache is an in-memory cache to account for existing image layers
+// It is used primarily by imagec when coordinating layer downloads
+// The cache is initially hydrated by way of the image cache at startup
```

Comment:
@caglar10ur I agree that getting rid of the cache would be a good thing, but if we use the fs semantics, I don't think that would allow another client to detect that a layer has already been downloaded. Example:

Client A pulls tomcat. He gets the lock and starts putting all the layers in flight. The lock releases and the layers are downloading.

Client B comes along and pulls the same image. He gets the lock, calls out to the PL and learns what images he needs to download (say all of them, since none have been written to the PL yet). At this point, the file locks will indicate which layers are in flight (good).

But what if Client A completes downloading some particular layer, writes to the PL and releases the file lock before Client B attempts to obtain the file lock?

Client B is not aware of this, obtains the file lock and repeats the download for that layer, eventually getting an "Already exists" error when attempting to write to the PL. Using the cache allows for Client B to know that a layer has already been downloaded. The existence of the layer in the cache is the indicating factor that Client B should not attempt a download, but the state (true/false) of the cached layer also indicates whether or not Client B should track that layer's progress during a download. Is there a way to accomplish this using the fs semantics?